Mathlib/Dynamics/SymbolicDynamics/Basic.lean

```diff
@@ -0,0 +1,316 @@
+/-
+Copyright (c) 2025 S. Gangloff. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: S. Gangloff
+-/
+
+import Mathlib.Topology.Basic
+import Mathlib.Topology.Constructions
+import Mathlib.Data.Finset.Basic
+import Mathlib.Logic.Equiv.Defs
+
+/-!
+# Symbolic dynamics on groups
+
+This file develops a minimal API for symbolic dynamics over an arbitrary group `G`.
+The ambient space is the full shift
+`FullShift A G := G → A` (an `abbrev`), hence it inherits the product topology
+from the Π-type. We define the right-translation action, cylinders, finite patterns,
+their occurrences, forbidden sets, and subshifts (closed, shift-invariant subsets).
+Basic topological statements (e.g. cylinders are clopen, occurrence sets are clopen,
+forbidden sets are closed) are proved under discreteness assumptions on the alphabet.
+
+The file is group-generic. Geometry specific to `ℤ^d` (boxes/cubes and the
+box-based entropy) is kept in a separate specialization.
+
+## Main definitions
+
+* `FullShift A G := G → A`.
+* `shift g x` — right translation: `(shift g x) h = x (h * g)`.
+* `cylinder U x` — configurations agreeing with `x` on a finite set `U ⊆ G`.
+* `Pattern A G` — finite support together with values on that support.
+* `Pattern.occursIn p x g` — occurrence of `p` in `x` at translate `g`.
+* `forbids F` — configurations avoiding every pattern in `F`.
+* `Subshift A G` — closed, shift-invariant subsets of the full shift.
+
+## Conventions
+
+We use a **right** action of `G` on configurations:
+`(shift g x) h = x (h * g)`. In additive notation (e.g. for `ℤ^d`) this is
+`(shift v x) u = x (u + v)`.
+
+## Implementation notes
+
+* Since `FullShift A G` is an `abbrev` for `G → A`, instances such as
+  `TopologicalSpace (FullShift A G)` and `Inhabited (FullShift A G)` are inherited
+  automatically from the Π-type; no explicit instances are declared here.
+* Openness/closedness results for cylinders and occurrence sets use
+  `[DiscreteTopology A]`. The closedness proofs that enumerate values additionally
+  require `[Fintype A]`, `[DecidableEq A]`, and `[DecidableEq G]` (for `Finset` manipulations
+  and `Function.update`).
+-/
+
+-- TODO:
+-- In entropy, match the namespaces of basic2 ?
+-- 3. Rewrite documentation
+-- 4. CI checks
+-- 5. Rewrite the PR description ? Make comments ?
+
+noncomputable section
+open Set Topology
+
+namespace SymbolicDynamics
+
+variable {A : Type*} [Fintype A] [Fintype A] [DecidableEq A] [Inhabited A]
+variable {G : Type*}
+variable [TopologicalSpace A] [DiscreteTopology A]
+variable [Group G] [DecidableEq G]
+
+/-! ## Full shift and shift action -/
+
+
+/-- Right-translation shift: `(shift g x)(h) = x (h * g)`. -/
+def shift (g : G) (x : G → A) : G → A :=
```

Comment:
Down the road, you will want to apply this mainly to additive groups, like Z or Z^d. Currently, your framework is only done for multiplicative groups. It is probably a good idea to do it simultaneously for multiplicative and additive groups. We have a device for this, called `to_additive`: do everything in the multiplicative setting, but tag definitions and results for `to_additive` and it will create automatically also the additive version.

Some care should be taken with names. For instance, you could use `mulShift` for the shift in a multiplicative context, and `addShift` in an additive context (if you tag a definition called `mulShift` with `to_additive`, it should pick automatically the name `addShift` for the additivized version)